### DISCUSSION TITLE
Speed differences between running directly on llama.cpp and via ooba texgen

### DISCUSSION THREAD

**Discussion by Wolvensdale** [oobabooga/text-generation-webui#830] [Q&A] [answered]
Hi, i've been running various models on alpaca, llama, and gpt4all repos, and they are quite fast. After instruct command it only take maybe 2 to 3 second for the models to start writing the replies.

As i know here, ooba also already integrate llama.cpp and support ggml. But why is the speed of text generation here is really slow? it needs 100 seconds at least to start generating response. Is this normal or something need to be tweaked?

Thanks in advance.

**Comment by oobabooga** [comment]
Try the new updated implementation, it might be better

**Reply by Wolvensdale** [reply]
For some reason I got this error instead after updating.

CUDA SETUP: Loading binary E:\Oobaboga\oobabooga\installer_files\env\lib\site-packages\bitsandbytes\libbitsandbytes_cuda116.dll...
Loading vicuna-13b-ggml...
Traceback (most recent call last):
  File "E:\Oobaboga\oobabooga\text-generation-webui\server.py", line 308, in <module>
    shared.model, shared.tokenizer = load_model(shared.model_name)
  File "E:\Oobaboga\oobabooga\text-generation-webui\modules\models.py", line 106, in load_model
    from modules.llamacpp_model_alternative import LlamaCppModel
  File "E:\Oobaboga\oobabooga\text-generation-webui\modules\llamacpp_model_alternative.py", line 9, in <module>
    from llama_cpp import Llama
ModuleNotFoundError: No module named 'llama_cpp'
Press any key to continue . . .

Oh and about the speed, is it because i choose to run it on GPU mode?
Probably i should reinstall it and using option B (CPU) instead since llama.cpp are using CPU?

**Reply by mcmonkey4eva** [reply, marked as answer]
@Wolvensdale check https://github.com/oobabooga/text-generation-webui/wiki/llama.cpp-models and make sure you've run the `pip install` line on there. That error is telling you llamacpp (or at least an up-to-date version) is not installed on your system.
Make sure to run it within your venv too, which appears to be at `E:\Oobaboga\oobabooga\installer_files\env\`

So in full a line would be something like `E:\Oobaboga\oobabooga\installer_files\env\Scripts\pip install llama-cpp-python==0.1.23`
though note that line is based on some assumptions about your setup so it might be wrong.

---

LLaMA.cpp runs on CPU, non-llamacpp runs on GPU.

If you've used an installer and selected to not install CPU mode, then, yeah, that'd be why it didn't install CPU support automatically, and you can indeed try rerunning the installer with CPU selected as it may automate the steps I described above anyway.

Though if you selected GPU install because you have a good GPU and want to use it, run the webui with a non-ggml model and enjoy the speed of your GPU.

**Reply by Wolvensdale** [reply]
Thanks, managed to run it now after update today.

About the speed, in my case its because im using too many context in the character.
When im using simple one like "this is a chatbot that will answer and explain everything".
It's becoming fast, nearly as fast as using llama.cpp directly.

**Comment by powercore2000** [comment]
I'm having the same issue, running .ggml files with llama.cpp in the UI returns 2 tokens/second at max, it causes a long time delay, and response time degrades as context gets larger. However, but running it on llama.cpp directly is far faster. 

I'm not sure what could be causing it, a bug with llama-cpp-python perhaps?

**Reply by Wolvensdale** [reply]
yeah larger context will make it really slow.
can only use a simple explanation for now in the character description.
Or use the default "assistant".

But again, probably due to lack of RAM.
How much RAM do you have? and what model are loaded in your case?

**Reply by powercore2000** [reply]
I have 32 GB of ram, and I tried running in assistant mode, but the ai only uses 5GB of ram, and 100% of my CPU for 2/tokens per second results.

These were run on 13b-vicuna-4bit-ggml model

**Reply by malixsys** [reply]
Works really fast in llama.cpp (mulitple tokens / second) but really slow in oobabooga (seconds / token)
Maybe a proper flag thing?

**Comment by freckletonj** [comment]
This is still an issue. 

`llama.cpp` + 7B model gives `20tok/s`

`llama.cpp` + 7B model + `oobabooga` gives like, `0.01tok/s`, or 1 tok per minute or so.
